crates/wasi-common/src/virtfs/pipe.rs

```diff
@@ -0,0 +1,412 @@
+                    | Rights::FD_FILESTAT_GET
+                    | Rights::POLL_FD_READWRITE,
+                Rights::empty(),
+            )),
+            writer,
+        }
+    }
+
+    /// Try to convert this `WritePipe<W>` back to the underlying `W` type.
+    ///
+    /// This will fail with `Err(self)` if multiple references to the underlying `W` exist.
+    pub fn try_into_inner(mut self) -> std::result::Result<W, Self> {
+        match Rc::try_unwrap(self.writer) {
+            Ok(rc) => Ok(RefCell::into_inner(rc)),
+            Err(writer) => {
+                self.writer = writer;
+                Err(self)
+            }
+        }
+    }
+}
+
+impl WritePipe<io::Cursor<Vec<u8>>> {
+    /// Create a new writable virtual pipe backed by a `Vec<u8>` buffer.
+    pub fn new_in_memory() -> Self {
+        Self::new(io::Cursor::new(vec![]))
+    }
+
+    /// Get a reference to the bytes contained in the underlying `Vec<u8>` buffer.
+    pub fn as_slice(&self) -> Ref<[u8]> {
+        Ref::map(self.writer.borrow(), |c| c.get_ref().as_slice())
+    }
+}
+
+impl<W: Write + Any> Handle for WritePipe<W> {
+    fn as_any(&self) -> &dyn Any {
+        self
+    }
+
+    fn try_clone(&self) -> io::Result<Box<dyn Handle>> {
+        Ok(Box::new(Self {
+            rights: self.rights.clone(),
+            writer: self.writer.clone(),
+        }))
+    }
+
+    fn get_file_type(&self) -> types::Filetype {
+        types::Filetype::Unknown
+    }
+
+    fn get_rights(&self) -> HandleRights {
+        self.rights.get()
+    }
+
+    fn set_rights(&self, rights: HandleRights) {
+        self.rights.set(rights)
+    }
+
+    fn advise(
+        &self,
+        _advice: types::Advice,
+        _offset: types::Filesize,
+        _len: types::Filesize,
+    ) -> Result<()> {
+        Err(Errno::Spipe)
+    }
+
+    fn allocate(&self, _offset: types::Filesize, _len: types::Filesize) -> Result<()> {
+        Err(Errno::Spipe)
+    }
+
+    fn fdstat_set_flags(&self, _fdflags: types::Fdflags) -> Result<()> {
+        // do nothing for now
+        Ok(())
+    }
+
+    fn filestat_get(&self) -> Result<types::Filestat> {
+        let stat = types::Filestat {
+            dev: 0,
+            ino: 0,
+            nlink: 0,
+            size: 0,
+            atim: 0,
+            ctim: 0,
+            mtim: 0,
+            filetype: self.get_file_type(),
+        };
+        Ok(stat)
+    }
+
+    fn filestat_set_size(&self, _st_size: types::Filesize) -> Result<()> {
+        Err(Errno::Spipe)
+    }
+
+    fn pwritev(&self, buf: &[io::IoSlice], offset: types::Filesize) -> Result<usize> {
+        if offset != 0 {
+            return Err(Errno::Spipe);
+        }
+        Ok(self.writer.borrow_mut().write_vectored(buf)?)
+    }
+
+    fn seek(&self, _offset: io::SeekFrom) -> Result<types::Filesize> {
+        Err(Errno::Spipe)
+    }
+
+    fn write_vectored(&self, iovs: &[io::IoSlice]) -> Result<usize> {
+        trace!("write_vectored(iovs={:?})", iovs);
```

Comment:
just noting: this is probably for consistency with the other virtfs `write_vectored` methods: https://github.com/bytecodealliance/wasmtime/blob/main/crates/wasi-common/src/virtfs.rs#L290 - I was pretty trace-y in all that and it might make sense to remove some of those too.

Reply:
Yes, I would prefer that are removed as well. Low priority, but nice to clean up.